# ETL HOMICIDIOS

In [2]:
import pandas as pd

In [3]:
hechos = pd.read_excel('data/homicidios.xlsx', sheet_name='HECHOS')

hechos.head()

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 XY (CABA)         pos x         pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  
3         MOTO-SD    MOTO         SD  
4  MOTO-PASAJEROS    MOTO  PASAJEROS  

[5 rows x 21 columns]

In [4]:
hechos.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Altura', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO'],
      dtype='object')

In [5]:
hechos.isnull().sum()

ID                         0
N_VICTIMAS                 0
FECHA                      0
AAAA                       0
MM                         0
DD                         0
HORA                       0
HH                         0
LUGAR_DEL_HECHO            0
TIPO_DE_CALLE              0
Calle                      1
Altura                   567
Cruce                    171
Dirección Normalizada      8
COMUNA                     0
XY (CABA)                  0
pos x                      0
pos y                      0
PARTICIPANTES              0
VICTIMA                    0
ACUSADO                    0
dtype: int64

In [6]:
hechos.shape

(696, 21)

El campo Altura tiene 567 datos nulos sobre el total de registros. Representa un valor significativo de tal manera que se elimina.

In [7]:
hechos= hechos.drop("Altura", axis=1)

Analizamos el resto de campos con nulos.

In [8]:
hechos[["Calle","Cruce","Dirección Normalizada"]][:10]

Calle                                Cruce  \
0   PIEDRA BUENA AV.  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1     PAZ, GRAL. AV.                  DE LOS CORRALES AV.   
2     ENTRE RIOS AV.                                  NaN   
3     LARRAZABAL AV.             VILLEGAS, CONRADO, GRAL.   
4       SAN JUAN AV.              SAENZ PE?A, LUIS, PRES.   
5  27 DE FEBRERO AV.                         ESCALADA AV.   
6             NOGOYA                 GONZALEZ, JOAQUIN V.   
7     PAZ, GRAL. AV.                  DE LOS CORRALES AV.   
8       BELGRANO AV.                IRIGOYEN, BERNARDO DE   
9     ENTRE RIOS AV.                                  NaN   

                               Dirección Normalizada  
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...  
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.  
2                                ENTRE RIOS AV. 2034  
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.  
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.  
5                   27 DE FEBRERO AV. y ESCALADA AV.  
6                      NOGOYA y GONZALEZ, JOAQUIN V.  
7               PAZ, GRAL. AV. y DE LOS CORRALES AV.  
8               BELGRANO AV. e IRIGOYEN, BERNARDO DE  
9                                ENTRE RIOS AV. 1366

Verificamos los duplicados

In [9]:
hechos.duplicated().sum()

0

In [10]:
hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Cruce                  525 non-null    object        
 12  Dirección Normalizada  688 non-null    object        
 13  COMUN

Según el diccionario de datos HH es	franja horaria entera.

In [11]:
hechos["HH"].unique()

array([4, 1, 7, 0, 5, 18, 19, 15, 11, 22, 16, 9, 23, 6, 10, 17, 12, 8, 20,
       21, 14, 3, 2, 13, 'SD'], dtype=object)

In [12]:
moda = hechos['HORA'].mode().iloc[0]

In [13]:
moda

datetime.time(14, 0)

In [14]:
hechos['HH'] = hechos['HH'].apply(lambda x: int(moda.hour) if x == "SD" else x)

Verifico el reemplazo por la moda.

In [15]:
hechos["HH"].unique()

array([ 4,  1,  7,  0,  5, 18, 19, 15, 11, 22, 16,  9, 23,  6, 10, 17, 12,
        8, 20, 21, 14,  3,  2, 13], dtype=int64)

In [17]:
hechos["COMUNA"].unique()

array([ 8,  9,  1, 11, 15,  4,  7, 12,  3, 13, 14, 10,  6,  2,  5,  0],
      dtype=int64)

Busco los nombres y barrios de las comunas. 

In [28]:
comunas= pd.read_excel("data/comunas_CABA.xlsx")

In [29]:
comunas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   COMUNA   15 non-null     int64 
 1   BARRIOS  15 non-null     object
dtypes: int64(1), object(1)
memory usage: 372.0+ bytes


In [30]:
hechos=hechos.merge(comunas, on="COMUNA", how="left")



In [31]:
hechos

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA  HH  \
0    2016-0001           1 2016-01-01  2016   1   1  04:00:00   4   
1    2016-0002           1 2016-01-02  2016   1   2  01:15:00   1   
2    2016-0003           1 2016-01-03  2016   1   3  07:00:00   7   
3    2016-0004           1 2016-01-10  2016   1  10  00:00:00   0   
4    2016-0005           1 2016-01-21  2016   1  21  05:20:00   5   
..         ...         ...        ...   ...  ..  ..       ...  ..   
691  2021-0093           1 2021-12-13  2021  12  13  17:10:00  17   
692  2021-0094           1 2021-12-20  2021  12  20  01:10:00   1   
693  2021-0095           1 2021-12-30  2021  12  30  00:43:00   0   
694  2021-0096           1 2021-12-15  2021  12  15  10:30:00  10   
695  2021-0097           1 2021-11-18  2021  11  18  06:10:00   6   

                               LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  \
0    AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...   
1            AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...   
2                           AV ENTRE RIOS 2034       AVENIDA  ...   
3        AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...   
4    AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...   
..                                         ...           ...  ...   
691                          AV. RIESTRA Y MOM       AVENIDA  ...   
692                    AU DELLEPIANE Y LACARRA     AUTOPISTA  ...   
693                        AV. GAONA Y TERRADA       AVENIDA  ...   
694                         AV. EVA PERON 4071       AVENIDA  ...   
695                    PADRE CARLOS MUJICA 709         CALLE  ...   

                                   Cruce  \
0    FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                    DE LOS CORRALES AV.   
2                                    NaN   
3               VILLEGAS, CONRADO, GRAL.   
4                SAENZ PE?A, LUIS, PRES.   
..                                   ...   
691                                  MOM   
692                          LACARRA AV.   
693                              TERRADA   
694                                  NaN   
695                                  NaN   

                                 Dirección Normalizada COMUNA  \
0    PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1                 PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                  ENTRE RIOS AV. 2034      1   
3            LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4              SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   
..                                                 ...    ...   
691                                  RIESTRA AV. y MOM      7   
692         DELLEPIANE, LUIS, TTE. GRAL. y LACARRA AV.      9   
693                                GAONA AV. y TERRADA     11   
694                                PERON, EVA AV. 4071      9   
695                            PADRE CARLOS MUGICA 709      1   

                                   XY (CABA)         pos x         pos y  \
0      Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1      Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2     Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3      Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4    Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   
..                                       ...           ...           ...   
691   Point (102728.60090138 98186.24929177)  -58.43353773  -34.64561636   
692    Point (99624.29795829 97569.69801131)  -58.46739825  -34.65117757   
693   Point (99116.45492358 101045.23284826)  -58.47293407  -34.61984745   
694    Point (99324.54463985 97676.26932409)  -58.47066794  -34.65021673   
695  Point (107664.16647795 104708.63962087)  -58.37976155  -34.58679619   

      PARTICIPANTES    VICTIMA    ACUSADO  \
0         MOTO-AUTO       MOTO       AUTO   
1    AUTO-PASAJEROS       AUTO  PASAJEROS

In [32]:
hechos.columns

Index(['ID', 'N_VICTIMAS', 'FECHA', 'AAAA', 'MM', 'DD', 'HORA', 'HH',
       'LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'Calle', 'Cruce',
       'Dirección Normalizada', 'COMUNA', 'XY (CABA)', 'pos x', 'pos y',
       'PARTICIPANTES', 'VICTIMA', 'ACUSADO', 'BARRIOS'],
      dtype='object')

In [33]:
hechos.columns = [x.capitalize() for x in hechos.columns]

In [34]:
hechos.columns

Index(['Id', 'N_victimas', 'Fecha', 'Aaaa', 'Mm', 'Dd', 'Hora', 'Hh',
       'Lugar_del_hecho', 'Tipo_de_calle', 'Calle', 'Cruce',
       'Dirección normalizada', 'Comuna', 'Xy (caba)', 'Pos x', 'Pos y',
       'Participantes', 'Victima', 'Acusado', 'Barrios'],
      dtype='object')

In [35]:
hechos["Comuna"] = hechos.apply(lambda x: "Comuna "+str(x["Comuna"]), axis=1)

# ETL VICTIMAS

In [37]:
victimas = pd.read_excel('data/Homicidios.xlsx', sheet_name='VICTIMAS')

In [38]:
victimas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 717 entries, 0 to 716
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID_hecho             717 non-null    object        
 1   FECHA                717 non-null    datetime64[ns]
 2   AAAA                 717 non-null    int64         
 3   MM                   717 non-null    int64         
 4   DD                   717 non-null    int64         
 5   ROL                  717 non-null    object        
 6   VICTIMA              717 non-null    object        
 7   SEXO                 717 non-null    object        
 8   EDAD                 717 non-null    object        
 9   FECHA_FALLECIMIENTO  717 non-null    object        
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 56.1+ KB


In [39]:
victimas.head(5)

ID_hecho      FECHA  AAAA  MM  DD        ROL VICTIMA       SEXO EDAD  \
0  2016-0001 2016-01-01  2016   1   1  CONDUCTOR    MOTO  MASCULINO   19   
1  2016-0002 2016-01-02  2016   1   2  CONDUCTOR    AUTO  MASCULINO   70   
2  2016-0003 2016-01-03  2016   1   3  CONDUCTOR    MOTO  MASCULINO   30   
3  2016-0004 2016-01-10  2016   1  10  CONDUCTOR    MOTO  MASCULINO   18   
4  2016-0005 2016-01-21  2016   1  21  CONDUCTOR    MOTO  MASCULINO   29   

   FECHA_FALLECIMIENTO  
0  2016-01-01 00:00:00  
1  2016-01-02 00:00:00  
2  2016-01-03 00:00:00  
3                   SD  
4  2016-02-01 00:00:00

In [40]:
victimas ["EDAD"] = pd.to_numeric(victimas["EDAD"], errors="coerce")
victimas ["EDAD"] = victimas["EDAD"].astype("Int64")

In [41]:
victimas ["EDAD"].unique()

<IntegerArray>
[  19,   70,   30,   18,   29,   22,   16,   59,   65,   34,   41,   50,   38,
   21,   52,   36,   20,   54, <NA>,   56,   24,   78,   79,   26,   57,   37,
   58,   23,   60,   42,   53,   51,   40,   87,   76,   75,   35,   80,   43,
   45,   67,   27,   55,   49,   81,   25,   33,   46,   83,   39,   28,    7,
   48,    4,   82,   32,   17,   47,   61,   10,   95,   73,   84,   66,   85,
    1,   15,   13,   77,   44,   31,   62,   74,   71,   11,   86,   69,   72,
   12,   63,   92,   68,   91,   64,    5,   88]
Length: 86, dtype: Int64

In [42]:
victimas["EDAD"] = victimas["EDAD"].fillna(0.0)

In [43]:
victimas ["EDAD"].unique()

<IntegerArray>
[19, 70, 30, 18, 29, 22, 16, 59, 65, 34, 41, 50, 38, 21, 52, 36, 20, 54,  0,
 56, 24, 78, 79, 26, 57, 37, 58, 23, 60, 42, 53, 51, 40, 87, 76, 75, 35, 80,
 43, 45, 67, 27, 55, 49, 81, 25, 33, 46, 83, 39, 28,  7, 48,  4, 82, 32, 17,
 47, 61, 10, 95, 73, 84, 66, 85,  1, 15, 13, 77, 44, 31, 62, 74, 71, 11, 86,
 69, 72, 12, 63, 92, 68, 91, 64,  5, 88]
Length: 86, dtype: Int64

In [44]:
victimas.isnull().sum()

ID_hecho               0
FECHA                  0
AAAA                   0
MM                     0
DD                     0
ROL                    0
VICTIMA                0
SEXO                   0
EDAD                   0
FECHA_FALLECIMIENTO    0
dtype: int64

In [45]:
victimas.columns = [x.capitalize() for x in victimas.columns]

Personalizar una columna para determinar el rango etario

In [47]:
# Definir los rangos etarios
bins = [0, 20, 40, 60, 80, 100]
labels = ['Menos de 20', ' Entre 21- 40', 'Entre 41-60', 'Entre 61-80', 'Más de 80']

# Utilizar pd.cut()
victimas['Rango_etario'] = pd.cut(victimas['Edad'], bins=bins, labels=labels, right=False)


In [48]:
victimas.head(5)

Id_hecho      Fecha  Aaaa  Mm  Dd        Rol Victima       Sexo  Edad  \
0  2016-0001 2016-01-01  2016   1   1  CONDUCTOR    MOTO  MASCULINO    19   
1  2016-0002 2016-01-02  2016   1   2  CONDUCTOR    AUTO  MASCULINO    70   
2  2016-0003 2016-01-03  2016   1   3  CONDUCTOR    MOTO  MASCULINO    30   
3  2016-0004 2016-01-10  2016   1  10  CONDUCTOR    MOTO  MASCULINO    18   
4  2016-0005 2016-01-21  2016   1  21  CONDUCTOR    MOTO  MASCULINO    29   

   Fecha_fallecimiento   Rango_etario  
0  2016-01-01 00:00:00    Menos de 20  
1  2016-01-02 00:00:00    Entre 61-80  
2  2016-01-03 00:00:00   Entre 21- 40  
3                   SD    Menos de 20  
4  2016-02-01 00:00:00   Entre 21- 40

Realizo un merge de los dos dataframes. Pero tengo que modificar el titulo del id de victimas.

In [53]:
victimas = victimas.rename(columns={'Id_hecho': 'Id'})

In [54]:
victimas.head(0)

Empty DataFrame
Columns: [Id, Fecha, Aaaa, Mm, Dd, Rol, Victima, Sexo, Edad, Fecha_fallecimiento, Rango_etario]
Index: []

In [50]:
hechos.head(0)

Empty DataFrame
Columns: [Id, N_victimas, Fecha, Aaaa, Mm, Dd, Hora, Hh, Lugar_del_hecho, Tipo_de_calle, Calle, Cruce, Dirección normalizada, Comuna, Xy (caba), Pos x, Pos y, Participantes, Victima, Acusado, Barrios]
Index: []

[0 rows x 21 columns]

In [55]:
victimas_hechos= victimas.merge(hechos, on="Id", how="left")

In [57]:
victimas_hechos.head(5)

Id    Fecha_x  Aaaa_x  Mm_x  Dd_x        Rol Victima_x       Sexo  \
0  2016-0001 2016-01-01    2016     1     1  CONDUCTOR      MOTO  MASCULINO   
1  2016-0002 2016-01-02    2016     1     2  CONDUCTOR      AUTO  MASCULINO   
2  2016-0003 2016-01-03    2016     1     3  CONDUCTOR      MOTO  MASCULINO   
3  2016-0004 2016-01-10    2016     1    10  CONDUCTOR      MOTO  MASCULINO   
4  2016-0005 2016-01-21    2016     1    21  CONDUCTOR      MOTO  MASCULINO   

   Edad  Fecha_fallecimiento  ...                                Cruce  \
0    19  2016-01-01 00:00:00  ...  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1    70  2016-01-02 00:00:00  ...                  DE LOS CORRALES AV.   
2    30  2016-01-03 00:00:00  ...                                  NaN   
3    18                   SD  ...             VILLEGAS, CONRADO, GRAL.   
4    29  2016-02-01 00:00:00  ...              SAENZ PE?A, LUIS, PRES.   

                               Dirección normalizada    Comuna  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...  Comuna 8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.  Comuna 9   
2                                ENTRE RIOS AV. 2034  Comuna 1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.  Comuna 8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.  Comuna 1   

                                 Xy (caba)         Pos x         Pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    Participantes  Victima_y    Acusado  \
0       MOTO-AUTO       MOTO       AUTO   
1  AUTO-PASAJEROS       AUTO  PASAJEROS   
2       MOTO-AUTO       MOTO       AUTO   
3         MOTO-SD       MOTO         SD   
4  MOTO-PASAJEROS       MOTO  PASAJEROS   

                                             Barrios  
0     VILLA LUGANO - VILLA RIACHUELO - VILLA SOLDATI  
1            LINIERS - MATADEROS - PARQUE AVELLANEDA  
2  CONSTITUCION - MONTSERRAT - PUERTO MADERO - RE...  
3     VILLA LUGANO - VILLA RIACHUELO - VILLA SOLDATI  
4  CONSTITUCION - MONTSERRAT - PUERTO MADERO - RE...  

[5 rows x 31 columns]

In [58]:
victimas_hechos.to_csv("data/unificado.csv", index=False)